src/server/cluster/outgoing_slot_migration.cc

```diff
@@ -153,72 +219,53 @@ bool OutgoingMigration::FinalyzeMigration(long attempt) {
-        return false;
-      }
+  if (auto resp = ReadRespReply(absl::GetFlag(FLAGS_slot_migration_connection_timeout_ms)); !resp) {
+    LOG(WARNING) << resp.error();
+    return false;
+  }
 
-      if (!CheckRespFirstTypes({RespExpr::INT64})) {
-        LOG(WARNING) << "Incorrect response type: "
-                     << facade::ToSV(LastResponseArgs().front().GetBuf());
-        return false;
-      }
-      attempt_res = get<int64_t>(LastResponseArgs().front().u);
-      if (attempt_res == kInvalidAttempt) {
-        return false;
-      }
-    } while (attempt_res != attempt);
+  if (!CheckRespFirstTypes({RespExpr::INT64})) {
+    LOG(WARNING) << "Incorrect response type: "
+                 << facade::ToSV(LastResponseArgs().front().GetBuf());
+    return false;
+  }
 
-    Finish();
+  const auto attempt_res = get<int64_t>(LastResponseArgs().front().u);
+  if (attempt_res == kInvalidAttempt) {
+    return false;
+  }
 
+  auto is_error = CheckFlowsForErrors();
+  Finish(is_error);
+  if (!is_error) {
     cf_->UpdateConfig(migration_info_.slot_ranges, false);
     VLOG(1) << "Config is updated for " << cf_->MyID();
-    return true;
-  } else {
-    // TODO implement connection issue error processing
   }
-  return false;
+  return true;
 }
 
-std::error_code OutgoingMigration::Start(ConnectionContext* cntx) {
-  VLOG(1) << "Starting outgoing migration";
-
-  state_.store(MigrationState::C_CONNECTING);
-
-  auto check_connection_error = [&cntx](error_code ec, const char* msg) -> error_code {
-    if (ec) {
-      cntx->SendError(absl::StrCat(msg, ec.message()));
-    }
-    return ec;
-  };
-
-  VLOG(1) << "Resolving host DNS";
-  error_code ec = ResolveHostDns();
-  RETURN_ON_ERR(check_connection_error(ec, "could not resolve host dns"));
-
-  VLOG(1) << "Connecting to source";
-  ec = ConnectAndAuth(absl::GetFlag(FLAGS_slot_migration_connection_timeout_ms) * 1ms, &cntx_);
-  RETURN_ON_ERR(check_connection_error(ec, "couldn't connect to source"));
-
-  VLOG(1) << "Migration initiating";
-  ResetParser(false);
-  auto cmd = absl::StrCat("DFLYMIGRATE INIT ", cf_->MyID(), " ", slot_migrations_.size());
-  for (const auto& s : migration_info_.slot_ranges) {
-    absl::StrAppend(&cmd, " ", s.start, " ", s.end);
+void OutgoingMigration::Start() {
+  VLOG(1) << "Resolving host DNS for outgoing migration";
+  if (error_code ec = ResolveHostDns(); ec) {
+    cntx_.ReportError(GenericError(ec, "Could not resolve host dns."));
+    return;
   }
-  RETURN_ON_ERR(SendCommandAndReadResponse(cmd));
-  LOG_IF(ERROR, !CheckRespIsSimpleReply("OK")) << facade::ToSV(LastResponseArgs().front().GetBuf());
 
   main_sync_fb_ = fb2::Fiber("outgoing_migration", &OutgoingMigration::SyncFb, this);
-
-  return {};
 }
 
+bool OutgoingMigration::CheckFlowsForErrors() {
+  for (const auto& flow : slot_migrations_) {
+    if (flow->GetError()) {
+      cntx_.ReportError(flow->GetError());
+      return true;
+    }
+  }
+  return false;
+}
```

Comment:
I guess you can just use a single context for the whole migration

Reply:
no, because different connection have different error processing

Reply:
but if a single one fails, you restart all, don't you?

Reply:
for flows yes, but for main connection I do reconnect